reconcile/dyn_traffic_director.py

```diff
@@ -0,0 +1,303 @@
+        # is flat (only one of each)
+        # TD <- Ruleset -> ResponsePool <- FailoverChain <- RecordSet <- Record
+        for ruleset in td.rulesets:
+            for pool in ruleset.response_pools:
+                for chain in pool.rs_chains:
+                    for recordset in chain.record_sets:
+                        for record in recordset.records:
+                            rdata = record.rdata()
+                            if 'cname_rdata' not in rdata:
+                                raise UnsupportedRecordType(
+                                    f'record type {type(record)} is not '
+                                    f'supported: {record}'
+                                )
+
+                            cname_rdata = record.rdata()['cname_rdata']
+                            if 'cname' not in cname_rdata:
+                                raise UnsupportedRecordType(
+                                    f'record missing a cname field: {record}'
+                                )
+
+                            records.append({
+                                'hostname':
+                                # strip trailing dot added by Dyn
+                                cname_rdata['cname'].rstrip('.'),
+                                'weight':
+                                int(cname_rdata.get('weight', DEFAULT_WEIGHT)),
+                            })
+
+        # need to be sorted for comparison later
+        records = sorted(records, key=lambda d: d['hostname'])
+
+        state['tds'][td_name] = {
+            'name': td_name,
+            'nodes': td_nodes,
+            'ttl': td_ttl,
+            'records': records,
+        }
+
+    return state
+
+
+def fetch_desired_state() -> dict:
+    dyn_tds = queries.get_dyn_traffic_directors()
+
+    state: dict = {
+        'tds': {},
+    }
+
+    for td in dyn_tds:
+        td_name: str = td['name']
+        td_records: List[Dict] = td.get('records', [])
+        td_ttl: int = td.get('ttl', DEFAULT_TD_TTL)
+
+        records: List[Dict[str, Union[str, int]]] = []
+        for record in td_records:
+            if record['cluster']:
+                hostname = record['cluster']['elbFQDN']
+            elif record['hostname']:
+                hostname = record['hostname']
+            else:
+                raise InvalidRecord('either cluster or hostname must '
+                                    'be defined on a record')
+            records.append({
+                'hostname': hostname.rstrip('.'),
+                'weight': record['weight'],
+            })
+
+        # need to be sorted for comparison later
+        records = sorted(records, key=lambda d: d['hostname'])
+
+        state['tds'][td_name] = {
+            'name': td_name,
+            'nodes': [td_name],
+            'ttl': td_ttl,
+            'records': records,
+        }
+
+    return state
+
+
+def process_tds(current: dict, desired: dict,
```

Comment:
I have added more specific type hints in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/8af318cc0e3b8b85beaf8fb2efa0be97ba7ac6df

I think this is further against what you are suggesting here? I'd like to better understand why this can be an issue but here unless you think we really should not be doing this I would go ahead and leave it as is...

Reply:
I think there were two separate issues:

1. The other comment where I pointed out that using `Any` implicitly, like in `Dict` where that is really `Dict[Any, Any]` leads to needing to provide other types later in the method because there simply weren't hints to know that keys were strings.
2. Using abstract types for parameters is better because it makes methods more general.

Putting the two together we'd go from something like:

```
def my_method(data: Dict)
```
to

```
# Can use MutableMapping for cases where the object is being modified
def my_method(data: Mapping[str, Any]
```

That's just an example, but now the method can accept any Mapping (including a dictionary) and it has specific enough types that we know that the keys are `str`s, which might save us from providing other collection types explicitly. Sorry if providing this feedback separately made it seem contradicting (combined with it `Iterable` not having a type in one of the examples).

Reply:
Fixed style guide to hopefully avoid this confusion in the future: https://gitlab.cee.redhat.com/service/dev-guidelines/-/merge_requests/90

Reply:
Improved in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/1be499840fbdfedee8791e1932ba45fd1fc9fe40
Thanks for the additional explanation!